### PR TITLE
Feature/#114 construction tape

### LINKED ISSUE #561: Construction tape issues
When placing the tape these are some problems at the moment:

When there are stairs, it will remove them:
image

The corners render a bit weird:
image

### PR DESCRIPTION
**Pull request by nojo121**
Closes #561 #572 

# Changes proposed in this pull request:
-fixed placing
-made it craftable
-made it breakable
-added config if builder should place it
-made it passable
-added name in lang file

Review please

### CODE REVIEW

**Comment by Raycoms** [comment]
:shipit:  for me

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/578/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 1]
```diff
@@ -129,15 +129,22 @@ private void initBlock()
         setHardness(BLOCK_HARDNESS);
         setResistance(RESISTANCE);
         setLightOpacity(LIGHT_OPACITY);
-        setBlockUnbreakable();
     }
 
+    public boolean isReplaceable(IBlockAccess worldIn, BlockPos pos)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java]: MAJOR Add the "@Override" annotation above this method signature [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1161)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 2]
```diff
@@ -129,15 +129,22 @@ private void initBlock()
...
+    {
+        return true;
+    }
+    @Deprecated
+    public AxisAlignedBB getCollisionBoundingBox(IBlockState blockState, World worldIn, BlockPos pos)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
MAJOR Add the "@Override" annotation above this method signature [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1161)
INFO Do not forget to remove this deprecated code someday. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTapeCorner.java [hunk 3]
```diff
@@ -171,9 +171,17 @@ private void initBlock()
         setHardness(BLOCK_HARDNESS);
         setResistance(RESISTANCE);
         setLightOpacity(LIGHT_OPACITY);
-        setBlockUnbreakable();
     }
 
+    public boolean isReplaceable(IBlockAccess worldIn, BlockPos pos)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTapeCorner.java]: MAJOR Add the "@Override" annotation above this method signature [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1161)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTapeCorner.java [hunk 4]
```diff
@@ -171,9 +171,17 @@ private void initBlock()
...
+    {
+        return true;
+    }
+    @Deprecated
+    public AxisAlignedBB getCollisionBoundingBox(IBlockState blockState, World worldIn, BlockPos pos)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTapeCorner.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
MAJOR Add the "@Override" annotation above this method signature [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1161)
INFO Do not forget to remove this deprecated code someday. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/ConstructionTapeHelper.java [hunk 5]
```diff
@@ -100,95 +102,98 @@ public static void placeConstructionTape(@NotNull AbstractBuilding building, @No
 
     public static void placeConstructionTape(@NotNull WorkOrderBuild workOrder, @NotNull World world)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/ConstructionTapeHelper.java:109]: MAJOR The Cyclomatic Complexity of this method "placeConstructionTape" is 11 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/util/RecipeHandler.java [hunk 6]
```diff
@@ -28,6 +28,9 @@ private RecipeHandler()
      */
     public static void init(final boolean enableInDevelopmentFeatures, final boolean supplyChests)
     {
+        GameRegistry.addRecipe(new ItemStack(ModBlocks.blockConstructionTape, 16), "SWS", "S S", "S S", 'S', Items.STICK, 'W', new ItemStack(Blocks.WOOL, 1,4));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/RecipeHandler.java]: MINOR Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)
MINOR Assign this magic number 4 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/configuration/ConfigurationHandler.java [hunk 7]
```diff
@@ -38,6 +38,7 @@ public static synchronized void init(final File file)
         try
         {
             config.load();
+            builderPlaceConstructionTape = config.get(CATEGORY_GAMEPLAY, "placeConstructionTape", builderPlaceConstructionTape, "Should builder place construction tape").getBoolean();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/configuration/ConfigurationHandler.java]: MINOR Split this 183 characters long line (which is greater than 180 authorized). [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00103)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 12 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 8]
```diff
@@ -129,15 +129,27 @@ private void initBlock()
...
+    public boolean isReplaceable(IBlockAccess worldIn, BlockPos pos)
+    {
+        return true;
+    }
+    /**
+     * @deprecated (Remove this as soon as minecraft offers anything better).
+     */
+    @Override
+    @Deprecated
+    public AxisAlignedBB getCollisionBoundingBox(IBlockState blockState, World worldIn, BlockPos pos)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java:144]: INFO Do not forget to remove this deprecated code someday. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTapeCorner.java [hunk 9]
```diff
@@ -171,10 +171,23 @@ private void initBlock()
...
+    public boolean isReplaceable(IBlockAccess worldIn, BlockPos pos)
+    {
+        return true;
+    }
+    /**
+     * @deprecated (Remove this as soon as minecraft offers anything better).
+     */
+    @Override
+    @Deprecated
+    public AxisAlignedBB getCollisionBoundingBox(IBlockState blockState, World worldIn, BlockPos pos)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTapeCorner.java:186]: INFO Do not forget to remove this deprecated code someday. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/util/RecipeHandler.java [hunk 10]
```diff
@@ -28,6 +28,10 @@ private RecipeHandler()
      */
     public static void init(final boolean enableInDevelopmentFeatures, final boolean supplyChests)
     {
+        GameRegistry.addRecipe(new ItemStack(ModBlocks.blockConstructionTape, 16), "SWS", "S S", "S S",
+          'S', Items.STICK, 'W', new ItemStack(Blocks.WOOL, 1,4));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/RecipeHandler.java]: MINOR Assign this magic number 4 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/util/RecipeHandler.java [hunk 11]
```diff
@@ -28,6 +28,10 @@ private RecipeHandler()
      */
     public static void init(final boolean enableInDevelopmentFeatures, final boolean supplyChests)
     {
+        GameRegistry.addRecipe(new ItemStack(ModBlocks.blockConstructionTape, 16), "SWS", "S S", "S S",
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/RecipeHandler.java:32]: MINOR Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/578/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/ConstructionTapeHelper.java [hunk 12]
```diff
@@ -49,17 +50,18 @@ public static int checkIfPlaceable(@NotNull int x, @NotNull int y, @NotNull int
...
             final BlockPos blockMin1 = new BlockPos(x, newY-1, z);
-            if (world.getBlockState(block).isFullBlock() || world.getBlockState(block).getMaterial().isLiquid())
-                {newY = newY+1;}
-            else
-                if (world.getBlockState(blockMin1).isFullBlock() || world.getBlockState(blockMin1).getMaterial().isLiquid())
-                    {working = false;}
+            if (world.getBlockState(block).getMaterial().isReplaceable())
+                {
+                    if (world.getBlockState(blockMin1).getMaterial().isReplaceable())
+                        {newY = newY-1;}
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/ConstructionTapeHelper.java]: I'm sorry but this formats really weird :O
Can we change that to normal formatting?

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java [hunk 13]
```diff
@@ -951,7 +953,8 @@ protected boolean isPassable(@NotNull final IBlockState block)
         {
             if (block.getMaterial().blocksMovement())
             {
-                return block.getBlock() instanceof BlockDoor || block.getBlock() instanceof BlockFenceGate;
+                return block.getBlock() instanceof BlockDoor || block.getBlock() instanceof BlockFenceGate || block.getBlock() instanceof BlockConstructionTape
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java]: If you break it up, please break each || on a new line :)

File: src/main/java/com/minecolonies/coremod/util/RecipeHandler.java (hunk 10 above)
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/util/RecipeHandler.java]: we should have here the color constant to spell what this number means as a color

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java [hunk 14]
```diff
@@ -985,6 +988,11 @@ protected SurfaceType isWalkableSurface(@NotNull final IBlockState blockState)
             return SurfaceType.NOT_PASSABLE;
         }
 
+        if(block instanceof BlockConstructionTape || block instanceof BlockConstructionTapeCorner)
+        {
+            return SurfaceType.DROPABLE;
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java:995]: What does that mean? I am not in the code for that at the moment, sorry...

**Review by marvin-bitterlich — CHANGES_REQUESTED** [review]
Just some style remarks I would like to have fixed. Nothing mayor ;)

I'll do request changes so I can just hit approve once fixed

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java (hunk 14 above)
Comment (nojo121) [reply src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java:995]: I don't know. Ask @Raycoms he edited the pathfinding.

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java:995]: That means that they can't walk on top of it and they drop in.

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java [hunk 15]
```diff
@@ -951,7 +953,10 @@ protected boolean isPassable(@NotNull final IBlockState block)
         {
             if (block.getMaterial().blocksMovement())
             {
-                return block.getBlock() instanceof BlockDoor || block.getBlock() instanceof BlockFenceGate;
+                return block.getBlock() instanceof BlockDoor ||
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java]: MINOR '||' should be on a new line. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AOperator_Wrap)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java [hunk 16]
```diff
@@ -951,7 +953,10 @@ protected boolean isPassable(@NotNull final IBlockState block)
...
+                         block.getBlock() instanceof BlockFenceGate ||
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java]: MINOR '||' should be on a new line. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AOperator_Wrap)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java [hunk 17]
```diff
@@ -951,7 +953,10 @@ protected boolean isPassable(@NotNull final IBlockState block)
...
+                         block.getBlock() instanceof BlockConstructionTape ||
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java]: MINOR '||' should be on a new line. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AOperator_Wrap)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Review by marvin-bitterlich — APPROVED** [review]

File: src/main/java/com/minecolonies/coremod/util/RecipeHandler.java [hunk 18]
```diff
@@ -13,6 +13,7 @@
 public final class RecipeHandler
 {
 
+    private static int     YELLOW_WOOL = 4;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/RecipeHandler.java:16]: MINOR Rename this field "YELLOW_WOOL" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS3008)
MINOR Assign this magic number 4 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by pmardle** [comment]
Good stuff guys :)

**Pull request merged**